src/main/groovy/lang/NumberRange.java

```diff
@@ -0,0 +1,536 @@
+        }
+        reverse = areReversed(from, to);
+        Number tempFrom;
+        Number tempTo;
+        if (reverse) {
+            tempFrom = to;
+            tempTo = from;
+        } else {
+            tempFrom = from;
+            tempTo = to;
+        }
+        if (tempFrom instanceof Short) {
+            tempFrom = tempFrom.intValue();
+        } else if (tempFrom instanceof Float) {
+            tempFrom = tempFrom.doubleValue();
+        }
+        if (tempTo instanceof Short) {
+            tempTo = tempTo.intValue();
+        } else if (tempTo instanceof Float) {
+            tempTo = tempTo.doubleValue();
+        }
+
+        if (tempFrom instanceof Integer && tempTo instanceof Long) {
+            tempFrom = tempFrom.longValue();
+        } else if (tempTo instanceof Integer && tempFrom instanceof Long) {
+            tempTo = tempTo.longValue();
+        }
+
+        this.from = (Comparable) tempFrom;
+        this.to = (Comparable) tempTo;
+        this.stepSize = stepSize == null ? 1 : stepSize;
+        this.inclusive = inclusive;
+    }
+
+    /**
+     * For a NumberRange with step size 1, creates a new NumberRange with the same
+     * <code>from</code> and <code>to</code> as this NumberRange
+     * but with a step size of <code>stepSize</code>.
+     *
+     * @param stepSize the desired step size
+     * @return a new NumberRange
+     */
+    public <T extends Number & Comparable> NumberRange by(T stepSize) {
+        if (!Integer.valueOf(1).equals(this.stepSize)) {
+            throw new IllegalStateException("by only allowed on ranges with original stepSize = 1 but found " + this.stepSize);
+        }
+        return new NumberRange(comparableNumber(from), comparableNumber(to), stepSize, inclusive);
+    }
+
+    @SuppressWarnings("unchecked")
+    /* package private */ static <T extends Number & Comparable> T comparableNumber(Comparable c) {
+        return (T) c;
+    }
+
+    @SuppressWarnings("unchecked")
+    /* package private */ static <T extends Number & Comparable> T comparableNumber(Number n) {
+        return (T) n;
+    }
+
+    private static boolean areReversed(Number from, Number to) {
+        try {
+            return compareGreaterThan(from, to);
+        } catch (ClassCastException cce) {
+            throw new IllegalArgumentException("Unable to create range due to incompatible types: " + from.getClass().getSimpleName() + ".." + to.getClass().getSimpleName() + " (possible missing brackets around range?)", cce);
+        }
+    }
+
+    public boolean equals(Object that) {
+        return (that instanceof NumberRange) ? equals((NumberRange) that) : super.equals(that);
+    }
+
+    /**
+     * Compares an {@link NumberRange} to another {@link NumberRange}.
+     *
+     * @param that the object to check equality with
+     * @return <code>true</code> if the ranges are equal
+     */
+    public boolean equals(NumberRange that) {
```

Comment:
I also agree that the usefulness of the alternate implementations is worthy of discussion. Certainly those implementations will be ignored by the existing map classes and only become relevant for people wanting to write their own map-friendly map classes or for sorting/comparators. Also, whether they belong within NumberRange or in some Range utility or ListEquivalence class is worthy of consideration. Having said that, the implementations might at least better clarify the current behavior to some degree.